mlflow/gateway/providers/gemini.py

```diff
@@ -1,12 +1,241 @@
+import time
 from typing import Any
 
 from mlflow.gateway.config import GeminiConfig, RouteConfig
+from mlflow.gateway.exceptions import AIGatewayException
 from mlflow.gateway.providers.base import BaseProvider, ProviderAdapter
-from mlflow.gateway.providers.utils import send_request
-from mlflow.gateway.schemas import embeddings
+from mlflow.gateway.providers.utils import rename_payload_keys, send_request
+from mlflow.gateway.schemas import chat, completions, embeddings
+
+GENERATION_CONFIG_KEY_MAPPING = {
+    "stop": "stopSequences",
+    "n": "candidateCount",
+    "max_tokens": "maxOutputTokens",
+    "top_k": "topK",
+    "top_p": "topP",
+}
 
 
 class GeminiAdapter(ProviderAdapter):
+    @classmethod
+    def chat_to_model(cls, payload, config):
+        # Documentation: https://ai.google.dev/api/generate-content
+        # Example payload for the chat API.
+        #
+        # {
+        #     "contents": [
+        #         {
+        #             "role": "user",
+        #             "parts": [
+        #                 {
+        #                     "text": "System: You are an advanced AI Assistant"
+        #                 }
+        #             ]
+        #         },
+        #         {
+        #             "role": "user",
+        #             "parts": [
+        #                 {
+        #                     "text": "Please give the code for addition of two numbers"
+        #                 }
+        #             ]
+        #         }
+        #     ],
+        #     "generationConfig": {
+        #         "temperature": 0.1,
+        #         "topP": 1,
+        #         "stopSequences": ["\n"],
+        #         "candidateCount": 1,
+        #         "maxOutputTokens": 100,
+        #         "topK": 40
+        #     }
+        # }
+
+        for k1, k2 in GENERATION_CONFIG_KEY_MAPPING.items():
+            if k2 in payload:
+                raise AIGatewayException(
+                    status_code=422, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+
+        if "top_p" in payload and payload["top_p"] > 1:
+            raise AIGatewayException(
+                status_code=422, detail="top_p should be less than or equal to 1"
+            )
+
+        payload = rename_payload_keys(payload, GENERATION_CONFIG_KEY_MAPPING)
+
+        contents, system_message = [], None
+        for message in payload["messages"]:
+            role = message["role"]
+
+            if role == "assistant":
+                role = "model"
+            elif role == "system":
+                system_message = {"parts": [{"text": message["content"]}]}
```

Comment:
I think we need to handle multiple system contents. So we need to initialize `{"parts": []}` when seeing the first system message, and then append contents to parts.